File: frontend/data/challenges.ts
import { requestJson } from "./auth";
import { BetOutcome } from "./bets";

export type ChallengeStatus = "pending" | "accepted" | "declined" | "cancelled" | "settled";
export type ChallengeType = "match" | "custom";

export type FriendChallenge = {
  id: string;
  challengerAccountId: string;
  challengerDisplayName: string;
  opponentAccountId: string;
  opponentDisplayName: string;
  challengeType: ChallengeType;
  matchId: string | null;
  matchLabel: string | null;
  title: string | null;
  options: string[];
  // For "match" challenges this is a BetOutcome; for "custom" challenges it's whichever
  // option the challenger picked out of `options`.
  outcome: string;
  status: ChallengeStatus;
  createdAt: string;
  respondedAt: string | null;
  settledAt: string | null;
  result: string | null;
  winnerAccountId: string | null;
};

export type ChallengeList = {
  incoming: FriendChallenge[];
  outgoing: FriendChallenge[];
  active: FriendChallenge[];
  resolved: FriendChallenge[];
};

export async function listMyChallenges(): Promise<ChallengeList> {
  return requestJson<ChallengeList>("/challenges/mine");
}

export async function createMatchChallenge(
  opponentAccountId: string,
  matchId: string,
  outcome: BetOutcome
): Promise<FriendChallenge> {
  return requestJson<FriendChallenge>("/challenges", {
    method: "POST",
    body: JSON.stringify({ challengeType: "match", opponentAccountId, matchId, outcome }),
  });
}

export async function createCustomChallenge(
  opponentAccountId: string,
  title: string,
  options: string[],
  outcome: string
): Promise<FriendChallenge> {
  return requestJson<FriendChallenge>("/challenges", {
    method: "POST",
    body: JSON.stringify({ challengeType: "custom", opponentAccountId, title, options, outcome }),
  });
}

export async function acceptChallenge(challengeId: string): Promise<FriendChallenge> {
  return requestJson<FriendChallenge>(`/challenges/${encodeURIComponent(challengeId)}/accept`, {
    method: "POST",
  });
}

export async function declineChallenge(challengeId: string): Promise<FriendChallenge> {
  return requestJson<FriendChallenge>(`/challenges/${encodeURIComponent(challengeId)}/decline`, {
    method: "POST",
  });
}

export async function cancelChallenge(challengeId: string): Promise<FriendChallenge> {
  return requestJson<FriendChallenge>(`/challenges/${encodeURIComponent(challengeId)}/cancel`, {
    method: "POST",
  });
}

export async function resolveCustomChallenge(challengeId: string, result: string): Promise<FriendChallenge> {
  return requestJson<FriendChallenge>(`/challenges/${encodeURIComponent(challengeId)}/resolve`, {
    method: "POST",
    body: JSON.stringify({ result }),
  });
}

File: backend/bethany_mock/routers/challenges.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Body, Depends

from ..challenge_repository import (
    cancel_challenge,
    create_custom_challenge,
    create_match_challenge,
    list_challenges_for_account,
    resolve_custom_challenge,
    respond_challenge,
)
from ..session import require_session, require_verified_session

router = APIRouter()


@router.get("/challenges/mine")
def get_my_challenges(account_id: str = Depends(require_session)) -> dict[str, Any]:
    return list_challenges_for_account(account_id)


@router.post("/challenges", status_code=201)
def create_challenge(
    payload: dict[str, Any] = Body(default={}),
    account_id: str = Depends(require_verified_session),
) -> dict[str, Any]:
    challenge_type = str(payload.get("challengeType", "match"))
    if challenge_type == "custom":
        options = payload.get("options", [])
        return create_custom_challenge(
            account_id,
            str(payload.get("opponentAccountId", "")),
            str(payload.get("title", "")),
            [str(option) for option in options] if isinstance(options, list) else [],
            str(payload.get("outcome", "")),
        )
    return create_match_challenge(
        account_id,
        str(payload.get("opponentAccountId", "")),
        str(payload.get("matchId", "")),
        str(payload.get("outcome", "")),
    )


@router.post("/challenges/{challenge_id}/accept")
def accept_challenge(challenge_id: str, account_id: str = Depends(require_verified_session)) -> dict[str, Any]:
    return respond_challenge(account_id, challenge_id, accept=True)


@router.post("/challenges/{challenge_id}/decline")
def decline_challenge(challenge_id: str, account_id: str = Depends(require_session)) -> dict[str, Any]:
    return respond_challenge(account_id, challenge_id, accept=False)


@router.post("/challenges/{challenge_id}/cancel")
def cancel_challenge_route(challenge_id: str, account_id: str = Depends(require_session)) -> dict[str, Any]:
    return cancel_challenge(account_id, challenge_id)


@router.post("/challenges/{challenge_id}/resolve")
def resolve_challenge_route(
    challenge_id: str,
    payload: dict[str, Any] = Body(default={}),
    account_id: str = Depends(require_session),
) -> dict[str, Any]:
    return resolve_custom_challenge(account_id, challenge_id, str(payload.get("result", "")))
